openshift-knative/eventing-kafka
# [release-v0.24] SRVKE-936: Add dispatcher monitoring support

-  targets Serverless 1.18
- adds support for monitoring the kafka dispacther
- monitoring is updated when eventing CR observability part changes eg. user disables monitoring. Tested cases where user disabled monitoring and then enables it again via the Eventing CR.
- once this is approved we can make a patch although its a very big due to the vendor changes
- split into several commits to separate the vendor noise


To be used in conjunction with: https://github.com/openshift-knative/serverless-operator/pull/1204

To test:

1) Build controller image with `ko publish cmd/channel/consolidated/controller/main.go`
2) Use that in the S-O csv. 
3) Launch the S-O using the PR above.
4) Create a kafka channel and link to an apiserversource and observe metrics:
```
cat <<-EOF | kubectl apply -f -
---
apiVersion: messaging.knative.dev/v1beta1
kind: KafkaChannel
metadata:
  name: my-kafka-channel
spec:
  numPartitions: 3
  replicationFactor: 1
---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: events-sa
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: event-watcher
rules:
  - apiGroups:
      - ""
    resources:
      - events
    verbs:
      - get
      - list
      - watch
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: k8s-ra-event-watcher
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: event-watcher
subjects:
  - kind: ServiceAccount
    name: events-sa
    namespace: default
---
apiVersion: sources.knative.dev/v1
kind: ApiServerSource
metadata:
  name: testevents
spec:
  serviceAccountName: events-sa
  mode: Resource
  resources:
    - apiVersion: v1
      kind: Event
  sink:
    ref:
      apiVersion: messaging.knative.dev/v1beta1
      kind: KafkaChannel
      name: my-kafka-channel
---
apiVersion: messaging.knative.dev/v1
kind: Subscription
metadata:
  name: my-subscription
  namespace: default
spec:
  channel:
    apiVersion: messaging.knative.dev/v1beta1
    kind: KafkaChannel
    name: my-kafka-channel
  delivery:
    deadLetterSink:
      ref:
        apiVersion: serving.knative.dev/v1
        kind: Service
        name: event-display
  subscriber:
    ref:
      apiVersion: serving.knative.dev/v1
      kind: Service
      name: event-display
---
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: event-display
spec:
  template:
    spec:
      containers:
        - # This corresponds to
          # https://github.com/knative/eventing-contrib/tree/master/cmd/event_display/main.go
          image: quay.io/openshift-knative/knative-eventing-sources-event-display

EOF
```
Metrics shown in the OCP related dashboard: xxxx.dev.openshift.com/monitoring/dashboards/grafana-dashboard-definition-knative-eventing-channel

![image](https://user-images.githubusercontent.com/7945591/133603544-e808d9ab-b503-4bca-ab38-950a28f480df.png)
